src/nrnoc/init.cpp

```diff
@@ -297,9 +299,10 @@ void hoc_last_init(void)
 	Symbol *s;
 
 	hoc_register_var(scdoub, (DoubVec*)0, (VoidFunc*)0);
-	nrn_threads_create(1);
+	nrn_threads_create(1,0); //TODO - 2nd param?
```

Comment:
What's this?

Reply:
Ah, forgot to check that. `nrn_threads_create` takes two parameters, In C98 the former was valid.

Reply:
At least you did not set your locale to `de_CH` and use `,` in floating point numbers :)

Reply:
@nrnhines with master we have:
```gdb
(venv) bb-fvfzj1tzlywp:build savulesc$ lldb bin/nrniv
(lldb) target create "bin/nrniv"
Current executable set to '/Users/savulesc/Workspace/nrn2/build/bin/nrniv' (x86_64).
(lldb) b nrn_threads_create
Breakpoint 1: where = libnrniv.dylib`nrn_threads_create + 19 at multicore.c:491:6, address = 0x0000000000100513
(lldb) r
Process 27782 launched: '/Users/savulesc/Workspace/nrn2/build/bin/nrniv' (x86_64)
libnrniv.dylib was compiled with optimization - stepping may behave oddly; variables may not be available.
Process 27782 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x0000000100215513 libnrniv.dylib`nrn_threads_create(n=1, parallel=3444150) at multicore.c:491:6 [opt]
   488  void nrn_threads_create(int n, int parallel) {
   489          int i, j;
   490          NrnThread* nt;
-> 491          if (nrn_nthread != n) {
   492  /*printf("sizeof(NrnThread)=%d   sizeof(Memb_list)=%d\n", sizeof(NrnThread), sizeof(Memb_list));*/
   493                  threads_free_pthread();
   494                  nrn_threads_free();
Target 0: (nrniv) stopped.
(lldb) p parallel
(int) $0 = 3444150
``` 
I wonder if we should keep that big value, since the thread creation logic in `nrn_create_threads` is binary bound to `nrn_thread_parallel`:
```
 if (nrn_thread_parallel_ != parallel) {
        threads_free_pthread();
        if (parallel) {
            threads_create_pthread();
        }
    }
```
 
`nrn_threads_create` is called from main and can be called from the parallel context via `"nthread"` : 
![Screenshot 2020-11-10 at 09 58 51](https://user-images.githubusercontent.com/46521150/98652069-60dbeb00-233b-11eb-8f42-4e3828669863.png)

Reply:
@nrnhines since now we have to pass a value for the 2nd argument, I feel like it should be a variable that says we want to use threads. But is that valid in the context of  `hoc_last_init() -> nrn_threads_create` ?
```suggestion
	nrn_threads_create(1, nrn_use_threads);
```
cc: @pramodk

Reply:
A 0 for the second arg means serial threads. It was useful for debugging but perhaps has outlived its usefulness for that purpose. It was also useful in some cases for higher performance  via increased cache efficiency since a minimum netcon delay interval could be  many time steps and the integration for each "NrnThread" could be done for all those time steps before going onto the next NrnThread.  With regard to ```hoc_last_init``` calling ```nrn_threads_create(1)``` That just starts the launch with one thread. Because it is 1, it does not matter whether serial or parallel. But 0 second arg would be appropriate. Here is a case where it is a bad idea to have extern declarations in files instead of an include file. I had promised myself that in some future transformation, all extern would be eliminated from *.[c cpp] files.

Reply:
Ok, in that case I will keep 0 and remove the TODO. We will remove extern in the near future :muscle: